test/switchlevels.c

```diff
@@ -82,11 +76,18 @@ static int compress_chunk(PREFIX3(stream) *strm, int level, int size, int last)
         goto free_buf;
     }
 
-    strm->next_in = buf;
-    strm->avail_in = size;
+    strm->next_in = NULL;
+    strm->avail_in = 0;
```

Comment:
Why this change? Doesn't seem to be doing the same thing as before? Why is the buffer "stale"? It seems `read_all` should fill it..

Reply:
`buf` is freed at the end of this function. So when we enter it again, we hit UAF in case `deflateParams()` needs to write something.

However, if we only move the allocation, then `deflateParams()` will perform compression, leaving `deflate()` with nothing to do. Therefore: give `deflateParams()` empty input buffer and let it flush. And then give the real input buffer to `deflate()`.

Reply:
I see.. What about considering moving `strm->next_out` and `strm->avail_out` below `strm->next_in` and `strm->avail_out` to make it easier to read?

Reply:
Hm, isn't this already the case? Or do you mean the following?

```
     strm->next_in = buf;
     strm->avail_in = size;
     strm->next_out = buf + size;
     strm->avail_out = compsize;
```

This won't work, because `deflateParams()` would run with empty input buffer (good), but stale output buffer.

Reply:
I think adding the comment made reviewing the code a lot easier... I've been around since early days of zlib-ng but I still haven't memorized what all functions do in detail...